### PR TITLE
[inductor] for UserDefinedTritonKernels don't mark all inputs as mutating

### PR DESCRIPTION
**Pull request by ColinPeppler**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124425

Take this example:
```
def _mul2(x):
    y = torch.empty_like(x)
    mul2_kernel[(10,)](
        in_ptr0=x, out_ptr=y,
        n_elements=x.numel(), BLOCK_SIZE=1,
    )
    return y

def f(x):
    for _ in range(4):
        x = _mul2(x)
    return x + 1
```

Currently, the codegen will show up like this. Notice, how we allocate 5 buffers of the same size.
```
# Source Nodes: [triton_kernel_wrapper_mutation], Original ATen: []
buf0 = empty_strided_cuda((10, ), (1, ), torch.float32)
mul2_kernel_0.run(in_ptr0=arg0_1, out_ptr=reinterpret_tensor(buf0, (10, ), (1, ), 0) ...)

# Source Nodes: [triton_kernel_wrapper_mutation_1], Original ATen: []
buf4 = empty_strided_cuda((10, ), (1, ), torch.float32)
mul2_kernel_0.run(in_ptr0=reinterpret_tensor(buf0, (10, ), (1, ), 0), out_ptr=reinterpret_tensor(buf4, (10, ), (1, ), 0) ...)

# Source Nodes: [triton_kernel_wrapper_mutation_2], Original ATen: []
buf8 = empty_strided_cuda((10, ), (1, ), torch.float32)
mul2_kernel_0.run(in_ptr0=reinterpret_tensor(buf4, (10, ), (1, ), 0), out_ptr=reinterpret_tensor(buf8, (10, ), (1, ), 0) ...)

# Source Nodes: [triton_kernel_wrapper_mutation_3], Original ATen: []
buf12 = empty_strided_cuda((10, ), (1, ), torch.float32)
mul2_kernel_0.run(in_ptr0=reinterpret_tensor(buf8, (10, ), (1, ), 0), out_ptr=reinterpret_tensor(buf12, (10, ), (1, ), 0) ...)

# Source Nodes: [add], Original ATen: [aten.add]
buf16 = empty_strided_cuda((10, ), (1, ), torch.float32)
triton_poi_fused_add_0.run(buf12, buf16, 10, grid=grid(10), stream=stream0)...)
return (buf16, )
```

With this PR, we want to see this. Notice, how we only allocate 2 buffers this time. The other 3 buffers are re-used.
```
# Source Nodes: [triton_kernel_wrapper_mutation], Original ATen: []
buf0 = empty_strided_cuda((10, ), (1, ), torch.float32)
mul2_kernel_0.run(in_ptr0=arg0_1, out_ptr=reinterpret_tensor(buf0, (10, ), (1, ), 0), ...)
del arg0_1

# Source Nodes: [triton_kernel_wrapper_mutation_1], Original ATen: []
buf2 = empty_strided_cuda((10, ), (1, ), torch.float32)
mul2_kernel_0.run(in_ptr0=reinterpret_tensor(buf0, (10, ), (1, ), 0), out_ptr=reinterpret_tensor(buf2, (10, ), (1, ), 0) ...)

# Source Nodes: [triton_kernel_wrapper_mutation_2], Original ATen: []
buf4 = buf0; del buf0  # reuse
mul2_kernel_0.run(in_ptr0=reinterpret_tensor(buf2, (10, ), (1, ), 0), out_ptr=reinterpret_tensor(buf4, (10, ), (1, ), 0) ...)

# Source Nodes: [triton_kernel_wrapper_mutation_3], Original ATen: []
buf6 = buf2; del buf2  # reuse
mul2_kernel_0.run(in_ptr0=reinterpret_tensor(buf4, (10, ), (1, ), 0), out_ptr=reinterpret_tensor(buf6, (10, ), (1, ), 0) ...)
del buf4

# Source Nodes: [add], Original ATen: [aten.add]
buf8 = buf6; del buf6  # reuse
triton_poi_fused_add_0.run(buf8, 10, grid=grid(10), stream=stream0)
return (buf8, )
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @amjames @desertfire @chauhang

Differential Revision: [D56379307](https://our.internmc.facebook.com/intern/diff/D56379307)

### CODE REVIEW

**Comment by ColinPeppler** [comment]
Btw I think `test_triton_kernel_dependancies` fails because lowering is being passed empty `tensors_to_clone`.  So, I digged a little deeper and see that we refine `tensors_to_clone` in post-grad.

https://github.com/pytorch/pytorch/blob/e0792cf3d6e0542354a87f7548cc3c7890c4defd/torch/_inductor/fx_passes/reinplace.py#L505-L516

```
AOT AUTOGRAD GRAPH
 ===== Forward graph 0 =====
 /data/users/colinpeppler/pytorch/torch/fx/_lazy_graph_module.py class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: "f32[5]", arg1_1: "f32[5]"):
        # File: /data/users/colinpeppler/pytorch/test/inductor/test_triton_kernels.py:479 in call_triton, code: add_kernel_autotuned[grid](x, y, output, n_elements)
        triton_kernel_wrapper_functional_proxy = torch._higher_order_ops.triton_kernel_wrap.triton_kernel_wrapper_functional(kernel_idx = 0, constant_args_idx = 0, grid = [(1, 1, 1)], kwargs = {'in_ptr0': arg0_1, 'in_ptr1': arg1_1, 'out_ptr': full}, tensors_to_clone = ['out_ptr']);  arg0_1 = full = None
        getitem: "f32[5]" = triton_kernel_wrapper_functional_proxy['out_ptr'];  triton_kernel_wrapper_functional_proxy = None
        
        # File: /data/users/colinpeppler/pytorch/test/inductor/test_triton_kernels.py:481 in call_triton, code: add_kernel_autotuned[grid](output, y, output2, n_elements)
        triton_kernel_wrapper_functional_proxy_1 = torch._higher_order_ops.triton_kernel_wrap.triton_kernel_wrapper_functional(kernel_idx = 0, constant_args_idx = 1, grid = [(1, 1, 1)], kwargs = {'in_ptr0': getitem, 'in_ptr1': arg1_1, 'out_ptr': full_1}, tensors_to_clone = ['out_ptr']);  getitem = arg1_1 = full_1 = None

 ===== AFTER POST GRAD =====
 /data/users/colinpeppler/pytorch/torch/fx/_lazy_graph_module.py class <lambda>(torch.nn.Module):
        # File: /data/users/colinpeppler/pytorch/test/inductor/test_triton_kernels.py:479 in call_triton, code: add_kernel_autotuned[grid](x, y, output, n_elements)
        triton_kernel_wrapper_functional_proxy = torch._higher_order_ops.triton_kernel_wrap.triton_kernel_wrapper_functional(kernel_idx = 0, constant_args_idx = 0, grid = [(1, 1, 1)], kwargs = {'in_ptr0': arg0_1, 'in_ptr1': arg1_1, 'out_ptr': full_default}, tensors_to_clone = []);  arg0_1 = None


        
        # File: /data/users/colinpeppler/pytorch/test/inductor/test_triton_kernels.py:481 in call_triton, code: add_kernel_autotuned[grid](output, y, output2, n_elements)
        triton_kernel_wrapper_functional_proxy_1 = torch._higher_order_ops.triton_kernel_wrap.triton_kernel_wrapper_functional(kernel_idx = 0, constant_args_idx = 1, grid = [(1, 1, 1)], kwargs = {'in_ptr0': full_default, 'in_ptr1': arg1_1, 'out_ptr': full_default_1}, tensors_to_clone = []);  full_default = arg1_1 = None
```

**Comment by ColinPeppler** [comment]
hey @oulgen, I'm having trouble thinking about a clean way to remember what tensors to mark as mutating. Because `tensors_to_clone` gets wiped in the reinplace post-grad pass, we have no clue which arg is mutating at lowering.

File: test/inductor/test_triton_kernels.py [hunk 1]
```diff
@@ -787,6 +787,38 @@ def f(x):
...
+                # but at some point we should start re-using buffers not allocate new buffers.
+                x = _mul2(x)
+            return x + 1
+
+        x = torch.randn(10, device="cuda", dtype=torch.float32)
+        eager_out = f(x)
+        compiled_out, (code,) = run_and_get_code(torch.compile(f), x)
+        self.assertEqual(compiled_out, eager_out)
+
+        # Check that we're re-using any unused memory buffers.
```
Comment (aakhundov) [inline test/inductor/test_triton_kernels.py:816]: Would it be more explicit to check that "# reuse" is in the code as many times as expected? Maybe in addition to the check we already have.

File: torch/_inductor/lowering.py [hunk 2]
```diff
@@ -5574,14 +5574,20 @@ def inner_fn(idx):
 
 
 @register_lowering(triton_kernel_wrapper_mutation)
-def triton_kernel_wrap_(*, kernel_idx, constant_args_idx, grid, kwargs):
+def triton_kernel_wrap_(
+    *, kernel_idx, constant_args_idx, grid, kwargs, tensors_to_clone=None
```
Comment (aakhundov) [inline torch/_inductor/lowering.py:5578]: Seems that you've updated the only call site of `triton_kernel_wrap_`. If so, let's make `tensors_to_clone` mandatory?

File: torch/_inductor/lowering.py [hunk 3]
```diff
@@ -5574,14 +5574,20 @@ def inner_fn(idx):
...
+):
     from torch._higher_order_ops.triton_kernel_wrap import kernel_side_table
 
+    if tensors_to_clone is None:
+        tensors_to_clone = []
```
Comment (aakhundov) [inline torch/_inductor/lowering.py:5583]: ...and then won't need this.

File: torch/_inductor/ir.py [hunk 4]
```diff
@@ -4489,9 +4489,12 @@ def __init__(self, *, kernel_idx, grid, kernel_args):
             arg for arg in kernel.arg_names if arg in kernel_args
         ]
 
-        mark_node_as_mutating(
-            self, *[a for a in kernel_args.values() if isinstance(a, TensorBox)]
-        )
+        mutable_args = [
+            arg
+            for key, arg in kernel_args.items()
```
Comment (oulgen) [inline torch/_inductor/ir.py:4494]: iterate over `tensors_to_clone` since key from that is guaranteed to be in kernel_args

**Comment by aakhundov** [comment]
> hey @oulgen, I'm having trouble thinking about a clean way to remember what tensors to mark as mutating. Because tensors_to_clone gets wiped in the reinplace post-grad pass, we have no clue which arg is mutating at lowering.

@ColinPeppler should we add another list, say `mutated_tensors`, which will be populated the same way as `tensors_to_clone` in the analysis pass, but won't be touched in the reinplacement pass? Then we could use `mutated_tensors` in the `UserDefinedTritonKernel.__init__` to decide what to make as mutated.

File: torch/_inductor/lowering.py (hunk 2 above)
Comment (oulgen) [reply torch/_inductor/lowering.py:5578]: This one matches the arguments of `triton_kernel_wrapper_mutation ` which does not need `tensors_to_clone`, so he does need that to be optional

**Review by oulgen — CHANGES_REQUESTED** [review]
I don't think this is gonna work. Tensors to clone is specifically for reinplacing logic, not for input/output tracking. For input/output, tracking perhaps memoize the analysis pass and call it again on the kernel?

File: test/inductor/test_triton_kernels.py (hunk 1 above)
Comment (ColinPeppler) [reply test/inductor/test_triton_kernels.py:816]: good call, that caught some weird case where we had a redundant back2back `reuse`. I have a fix for that but I'm thinking it's better to put that in a separate PR before this one.

**Review by oulgen — CHANGES_REQUESTED** [review]
Please do not use the side table for this. Why don't you just memoize the result of analysis and call again?

**Comment by oulgen** [comment]
> oh i see, let me try it that way

Just call the analysis pass the second time, and we can fix it if it becomes a perf issue

File: torch/_inductor/codegen/wrapper.py [hunk 5]
```diff
@@ -1497,6 +1497,7 @@ def can_reuse(self, input_buffer, output_buffer=None):
             or name in V.graph.graph_inputs
             or name in V.graph.constants
             or name in V.graph.never_reuse_buffers
+            or name not in self.allocated
```
Comment (ColinPeppler) [inline torch/_inductor/codegen/wrapper.py:1500]: had this weird case where we had back-to-back re-uses

```
buf11 = buf6; del buf6
buf12 = buf11; del buf11
```

File: torch/_inductor/ir.py [hunk 6]
```diff
@@ -4654,18 +4657,37 @@ def __init__(self, *, kernel_idx, grid, kernel_args):
...
             arg for arg in kernel.arg_names if arg in kernel_args
         ]
 
-        mark_node_as_mutating(
-            self, *[a for a in kernel_args.values() if isinstance(a, TensorBox)]
-        )
+
+        from torch._higher_order_ops.triton_kernel_wrap import identify_mutated_tensors
+
+        def _to_real_values(v):
```
Comment (oulgen) [inline torch/_inductor/ir.py:4669]: we already do this type of conversation in identify_mutated_tensors. From FakeTensor to real tensor, why dont you just add TensorBox there?

File: torch/_inductor/codegen/wrapper.py (hunk 5 above)
Comment (ColinPeppler) [reply torch/_inductor/codegen/wrapper.py:1500]: found out it was because we had ` V.kernel.inplace_update_buffers[buf12] = buf11`, although `buf11` is a `NopKernelSchedulerNode(MutationOutput)`.

That happens here.
https://github.com/pytorch/pytorch/blob/0feab7d6c3c7da284a86d6a0c7d0f0c84cff577e/torch/_inductor/scheduler.py#L404-L406

File: torch/_inductor/ir.py (hunk 6 above)
Comment (ColinPeppler) [reply torch/_inductor/ir.py:4669]: yeah not sure why I was afraid to add it there

File: torch/_higher_order_ops/triton_kernel_wrap.py [hunk 7]
```diff
@@ -137,9 +138,9 @@ def generate_ttir(kernel, kwargs):
     ordered_args: Dict[str, Any] = {}
     for name in kernel.arg_names:
         a = kwargs[name]
-        if isinstance(a, (torch.SymInt, torch.SymFloat, torch.SymBool)):
+        if isinstance(a, (torch.SymInt, torch.SymFloat, torch.SymBool, Expr)):
```
Comment (oulgen) [inline torch/_higher_order_ops/triton_kernel_wrap.py:141]:
```suggestion
        if isinstance(a, (torch.SymInt, torch.SymFloat, torch.SymBool, sympy.Expr)):
```

**Review by oulgen — APPROVED** [review]

**Comment by ColinPeppler** [comment]
@ColinPeppler has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56379307).

**Comment by ColinPeppler** [comment]
@ColinPeppler has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56379307).

**Comment by ColinPeppler** [comment]
lowered the expected # of memory accesses in test_perf.py

File: torch/_higher_order_ops/triton_kernel_wrap.py [hunk 8]
```diff
@@ -111,6 +111,7 @@ def generate_ttir(kernel, kwargs):
     """
     Uses Triton's internal code generation to create TTIR
     """
+    import sympy
```
Comment (oulgen) [inline torch/_higher_order_ops/triton_kernel_wrap.py:114]: we dont need this anymore?

Comment (ColinPeppler) [reply torch/_higher_order_ops/triton_kernel_wrap.py:114]: need it for sympy.Expr

File: torch/_inductor/codegen/wrapper.py (hunk 5 above)
Comment (ColinPeppler) [reply torch/_inductor/codegen/wrapper.py]: I can't add this line, some buffers are not allocated but are "reused". Like the buffer for a `ir.WhileLoop`. It can store subgraph output buffers, and eventually be re-used by a buffer in the parent graph.

https://github.com/pytorch/pytorch/blob/a32eac345f5d7862246c397d0a32b5e0cbca77ff/torch/_inductor/codegen/wrapper.py#L1585-L1594

So, instead we skip inserting into `inplace_update_buffers` if `isinstance(input_node, NopKernelSchedulerNode)`.

**Comment by ColinPeppler** [comment]
@ColinPeppler has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56379307).
